KIFUIViewTestActor

RoyalPineapple opened the pull request:
New KIFUIViewTestActor as outlined in https://github.com/kif-framework/KIF/issues/555
This actor decouples finding an element from acting upon the element.

This also allows you to define an actor for a specific view description, allowing you to reference the same view multiple times without worrying about stringly-typed errors. 
(see https://github.com/RoyalPineapple/KIF/blob/ca1b2baa5189e133c6aa50c6cbd86a1857f606c0/KIF%20Tests/PickerTests_ViewTestActor.m) for an example

This change will increase maintainiblity of the kif framework as well as our individual tests.

TBonnin commented on Classes/KIFUIViewTestActor.m:
```diff
@@ -0,0 +1,396 @@
+
+- (instancetype)usingExpectedClass:(Class)expectedClass;
+{
+    return [self usingPredicate:[NSPredicate predicateWithFormat:@"class MATCHES %@", expectedClass]];
```
Should not it be 
```
return [self usingPredicate:[NSPredicate predicateWithBlock:^BOOL(id evaluatedObject, NSDictionary *bindings) {
    return [evaluatedObject isKindOfClass:expectedClass];
}]];
```

RoyalPineapple replied:
Block predicates create an issue with our error logging as their `description` isn't very usefull.  You're absolutely right though, this is incorrect and wont work with subclasses, etc.

Thinking about it more, I dont think we should support this at all. I had initially implemented it to debug an issue where KIF was finding the wrong element, but the whole concept of a class is an implementation detail and  isn't user-facing, so it really doesnt belong. If needed anyone can build a class predicate and pass it into `usingPredicate:`

TBonnin commented on Classes/KIFUIViewTestActor.m:
```diff
@@ -0,0 +1,396 @@
+    int systemVersion = [UIDevice currentDevice].systemVersion.intValue;
+
+    if ([accessibilityLabel rangeOfString:@"\n"].location == NSNotFound || systemVersion == 6) {
+        return [self usingPredicate:[NSPredicate predicateWithFormat:@"accessibilityLabel MATCHES %@", accessibilityLabel]];
```
Using MATCHES implies the right hand expression is a regex. Why not use '==' instead? The right hand expression would always be a accessibilityLabel, accessibilityIdentifier, traits or value.

RoyalPineapple replied:
You are correct, this was an oversignt.

phatmann commented:
@RoyalPineapple, sorry to not respond sooner to this. I had a backlog of pull requests with crucial bug fixes I wanted to get through first.

As I said before, I like this change. Let me review it and think about the best way to roll it out. My thought right now is to place it in a separate subspec so people could review it, try it out, and tweak as needed. After that the next step would be to decide whether or not to deprecate the current API in favor of this one, which would then be the birth of KIF 4.0.

I encourage everyone to look at this pull request and weigh in with your thoughts and concerns.

In the meantime, @RoyalPineapple, can you please squash the 3 commits?

RoyalPineapple commented:
@phatmann I've squashed the commits and rebased off the upstream master. 
I also updated the UIViewTestActor tests to use category methods for commonly referenced views.

steipete commented:
As someone who just recently started to look into KIF, this PR is VERY interesting and a huge step forward. Definitely justifies a KIF 4.0 tag.

phatmann commented:
In order to merge this excellent pull request I want to gather more feedback and figure out the best way to package it up. I will be very busy until April 15 or so, so please be patient. You can be sure this is not being neglected! 

In the meantime, if anyone comes forward with a concrete packaging suggestion and is willing to implement it, I am all ears :-)

phatmann commented:
I am back now, and I will make a proposal for how to package this up this week or next week.

steipete commented:
Great! We've been adding tons of KIF tests so far and are happy to test a v4 branch.

mokagio commented:
I'd be happy to test version 4 too 👍

phatmann commented:
Lately there has been more and more confusion about the Identifier tests and the Conditional tests since they are separated out. In addition lots of folks want a new kind of test that works with either `accessibilityLabel` or `accessibilityIdentifier`. To avoid a combinatorial explosion, I believe we need to use a solution like this pull request. Let me review this pull request again, see if it feels good and suggest any possible changes. Once we have a plan we all agree on, we can go ahead and create KIF 4.0 based on it.

No matter what we decide, we need to make sure we have 100% backward compatibility with KIF 3.0 via a "legacy" layer.

steipete commented:
Fully agree that the current API is a bit confusing here. How about you create a new v4-branch based on this PR and then we figure out the details as we migrate the API to the new version. I don't really care about legacy support, that just clutters up autocomplete and makes the API larger.

Can I help out here? KIF's becoming really important for pspdfkit.com 's testing, so happy to put some time in.

RoyalPineapple commented:
Backwards compatability is important to me as we have a lot of tests that have yet to be updated to the new actor. 
As long as we keep the actors as seperate classes we should be able to move forward incrementally without polluting the autocomplete.

phatmann commented:
@RoyalPineapple, the idea of separate legacy classes using the existing names is a great idea. It will make backwards compatibility automatic for existing users, yet have no negative impact for new users.

phatmann commented:
I created a new v4 branch and merged this pull request into it. Let's move the discussion to #646.
